lib/view_model/set_up_2fa_viewmodel.dart

```diff
@@ -0,0 +1,159 @@
+// ignore_for_file: prefer_final_fields
+
+import 'package:cake_wallet/store/settings_store.dart';
+import 'package:cake_wallet/utils/totp_utils.dart' as Utils;
+import 'package:cake_wallet/view_model/auth_state.dart';
+import 'package:flutter/widgets.dart';
+import 'package:mobx/mobx.dart';
+import 'package:shared_preferences/shared_preferences.dart';
+
+import '../core/auth_service.dart';
+import '../core/execution_state.dart';
+import '../generated/i18n.dart';
+
+part 'set_up_2fa_viewmodel.g.dart';
+
+class Setup2FAViewModel = Setup2FAViewModelBase with _$Setup2FAViewModel;
+
+abstract class Setup2FAViewModelBase with Store {
+  final SettingsStore _settingsStore;
+  final AuthService _authService;
+  final SharedPreferences _sharedPreferences;
+
+  Setup2FAViewModelBase(this._settingsStore, this._sharedPreferences, this._authService)
+      : _failureCounter = 0,
+        enteredOTPCode = '',
+        state = InitialExecutionState() {
+    getRandomBase32SecretKey();
+    reaction((_) => state, _saveLastAuthTime);
+  }
+
+  static const maxFailedTrials = 3;
+  static const banTimeout = 180; // 3 minutes
+  final banTimeoutKey = S.current.auth_store_ban_timeout;
+
+  String get secretKey => _settingsStore.totpSecretKey;
+  String get deviceName => _settingsStore.deviceName;
+  String get totpVersionOneLink => _settingsStore.totpVersionOneLink;
+
+  @observable
+  ExecutionState state;
+
+  @observable
+  int _failureCounter;
+
+  @observable
+  String enteredOTPCode;
+
+  @computed
+  bool get useTOTP2FA => _settingsStore.useTOTP2FA;
+
+  void getRandomBase32SecretKey() {
+    final randomBase32Key = Utils.generateRandomBase32SecretKey(16);
+    setBase32SecretKey(randomBase32Key);
+  }
+
+  @action
+  void setUseTOTP2FA(bool value) {
+    _settingsStore.useTOTP2FA = value;
+  }
+
+  @action
+  void setBase32SecretKey(String value) {
```

Comment:
**Minor:**
can be private, to avoid any misuse

Reply:
Alright